moab_action_transform.ink

```diff
@@ -0,0 +1,132 @@
+
+# Default time delta between simulation steps (s)
+const DefaultTimeDelta = 0.045
+
+# Maximum distance per step in meters
+const MaxDistancePerStep = DefaultTimeDelta * MaxVelocity
+
+# Maximum plate angle in degrees
+const MaxPlateAngle = 22
+
+# State received from the simulator after each iteration
+type ObservableState {
+    # Ball X,Y position
+    ball_x: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    ball_y: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+
+    # Ball X,Y velocity
+    ball_vel_x: number<-MaxVelocity .. MaxVelocity>,
+    ball_vel_y: number<-MaxVelocity .. MaxVelocity>,
+}
+
+type BrainAction {
+    input_pitch: number<-MaxPlateAngle .. MaxPlateAngle>, # rotate about x-axis
+    input_roll: number<-MaxPlateAngle .. MaxPlateAngle>, # rotate about y-axis
+}
+
+# Action provided as output by policy and sent as
+# input to the simulator
+type SimAction {
+    # Range -1 to 1 is a scaled value that represents
+    # the full plate rotation range supported by the hardware.
+    input_pitch: number<-1 .. 1>, # rotate about x-axis
+    input_roll: number<-1 .. 1>, # rotate about y-axis
+}
+
+# Per-episode configuration that can be sent to the simulator.
+# All iterations within an episode will use the same configuration.
+type SimConfig {
+    # Model initial ball conditions
+    initial_x: number<-RadiusOfPlate .. RadiusOfPlate>, # in (m)
+    initial_y: number<-RadiusOfPlate .. RadiusOfPlate>,
+
+    # Model initial ball velocity conditions
+    initial_vel_x: number<-MaxInitialVelocity .. MaxInitialVelocity>, # in (m/s)
+    initial_vel_y: number<-MaxInitialVelocity .. MaxInitialVelocity>,
+
+    # Range -1 to 1 is a scaled value that represents
+    # the full plate rotation range supported by the hardware.
+    initial_pitch: number<-1 .. 1>,
+    initial_roll: number<-1 .. 1>,
+}
+
+function TransformAction(a: BrainAction): SimAction {
+    return {
+        input_pitch: a.input_pitch / MaxPlateAngle,
+        input_roll: a.input_roll / MaxPlateAngle,
+    }
+}
+
+# Define a concept graph with a single concept
+graph (input: ObservableState) {
+    concept MoveToCenter(input): BrainAction {
+        curriculum {
+            # The source of training for this concept is a simulator that
+            #  - can be configured for each episode using fields defined in SimConfig,
+            #  - accepts per-iteration actions defined in SimAction, and
+            #  - outputs states with the fields defined in SimState.
+            source simulator MoabSim(Action: SimAction, Config: SimConfig): ObservableState {
+                # Automatically launch the simulator with this
+                # registered package name.
+            }
+
+            training {
+                # Limit episodes to 250 iterations instead of the default 1000.
+                EpisodeIterationLimit: 250
+            }
+
+            action TransformAction
```

Comment:
Add a comment to explain what this is doing

Reply:
# Instruct the training/assessment session to use the action transform function
            # exporting a brain will still output the BrainAction type